### PR TITLE
Alternate key strategy oleg

### PR DESCRIPTION
**Pull request by phadej**
Follow-up to https://github.com/bos/aeson/pull/393

- [Original](http://oleg.fi/aeson-bench/datakinds.html)
- [without coerce](http://oleg.fi/aeson-bench/without-coerce.html)
- [with coerce](http://oleg.fi/aeson-bench/with-coerce.html)

Looks like that alternative approach is as fast, and is simpler. Good job @andrewthad !
Coerce approach gives a little speedup, don't know if it's worth the safety (common case in our codebase though, where the keys are `newtypes` of `Text`).

Didn't add dependency on `contravariant.

### CODE REVIEW

**Comment by phadej** [comment]
@andrewthad 
> I thought about this, but I feel like the dependency on `contravariant` probably not worth it when we could just provide a specialized version of the function instead.

`contravariant` is very light dependency, doesn't bring anything else except itself. But I leave it to maintainers to decide

**Comment by andrewthad** [comment]
@phadej I'm assuming you mean that `aeson` already transitively depends on everything that `contravariant` depends on. I haven't checked to see if that is the case, but if it is, then that would make me feel more comfortable with depending on it.

**Comment by phadej** [comment]
```diff
 packages:
 - '.'
 extra-deps:
+- StateVar-1.1.0.4
 - attoparsec-0.13.0.2
+- contravariant-1.4
 - dlist-0.7.1.2
 - fail-4.9.0.0
 - hashable-1.2.4.0
 - mtl-2.2.1
 - primitive-0.6.1.0
 - scientific-0.3.4.6
+- semigroups-0.18.1
+- stm-2.4.4.1
 - syb-0.6
 - tagged-0.8.4
 - text-1.2.2.1
+- transformers-compat-0.5.1.4
 - unordered-containers-0.2.7.0
 - vector-0.11.0.0
+- void-0.7.1
```

And with https://github.com/ekmett/contravariant/pull/27 the `void` and `semigroups` aren't pulled in.

**Comment by andrewthad** [comment]
I've opened up a PR against this PR that fixes two small things: https://github.com/phadej/aeson/pull/2

Thanks for doing the benchmarks to look into the performance issues. After spending a lot of time yesterday thinking about `unsafeCoerce` and inconsistent `Eq` and `Hashable` instances, I have ended up in favor of the `FromJSONKeyCoerce` data constructor. I think it is the best solution we have available. So I'm +1 on merging this in to master.

As a side note, in another branch, I played around with [using rewrite rules instead of unsafeCoerce](https://github.com/andrewthad/aeson/commit/7e8a740fd95d5791a457307740d421ad4d9e0540) to make decoding `HashMap`s with `Text` keys faster, but it ends up only working for `Text` and not for `newtype`s around `Text`, so the `unsafeCoerce` code ends up being necessary no matter regardless.

**Comment by andrewthad** [comment]
Also, @bergmark it would be good to hear your thoughts on the `contravariant` dependency.

**Comment by phadej** [comment]
- http://oleg.fi/aeson-bench/safer-coercible.html
- http://oleg.fi/aeson-bench/unsafe.html 

I'm not sure what happening with `Map` in benchmark in top version

File: Data/Aeson/Types.hs [hunk 1]
```diff
@@ -37,6 +37,14 @@ module Data.Aeson.Types
     , KeyValue(..)
     , modifyFailure
 
+    -- ** Keys for maps
+    , ToJSONKey(..)
+    , ToJSONKeyFunction(..)
+    , FromJSONKey(..)
+    , FromJSONKeyFunction(..)
+    , fromJSONKeyCoerce
+    , coerceFromJSONKeyFunction
```
Comment (phadej) [inline Data/Aeson/Types.hs]: These two functions aren't exported from `Data.Aeson` as they are quite special.

File: Data/Aeson/Types/Class.hs [hunk 2]
```diff
@@ -261,13 +276,50 @@ data ToJSONKeyFunction a
...
+
 data FromJSONKeyFunction a
-    = FromJSONKeyCoerce
+    = FromJSONKeyCoerce (CoerceText a)
     | FromJSONKeyText (Text -> a)
     | FromJSONKeyTextParser (Text -> Parser a)
     | FromJSONKeyValue (Value -> Parser a)
-  deriving (Functor)
 
+instance Functor FromJSONKeyFunction where
```
Comment (andrewthad) [inline Data/Aeson/Types/Class.hs]: Not a valid `Functor` instance since the following does not hold:

    fmap id x = x

Comment (phadej) [inline Data/Aeson/Types/Class.hs]: I'd say it's valid enough. Especially if we add `fmap id = coerceFromJSONKeyFunction` and `fmap coerce = coerceFromJSONKeyFunction` rewrite rules!

**Comment by andrewthad** [comment]
@phadej Concerning your two most recent benchmarks, my interpretation of them is that the safer one is faster than the unsafe one. Intuitively, that seems wrong. Am I reading these correctly?

**Comment by phadej** [comment]
They both take about of the same time. My machine isn't the best one to run the benchmarks (mail client etc. running). I posted them just to show that safer approach doesn't hurt the performance, if it's even faster - that's only good thing :) (might be that GHC can do something with `coerce`, which it cannot with `unsafeCoerce` - dunno about that).

**Comment by phadej** [comment]
Seems to be transient failure on travis

**Comment by bergmark** [comment]
> Also, @bergmark it would be good to hear your thoughts on the contravariant dependency.


Adding a dependency only for one instance sounds a bit excessive to me.

**Comment by phadej** [comment]
So is it good as it is now?

**Comment by andrewthad** [comment]
It looks good to me. The quirkiness of the `Functor` instance should probably be documented, but I'm planning on adding a more docs to these things in this PR after it's merged in, so that can wait for later.

**Comment by bergmark** [comment]
Just to make sure, do you guys think this is ready to be merged?

**Comment by phadej** [comment]
@bergmark yes

**Comment by andrewthad** [comment]
yes

**Comment by bergmark** [comment]
👏 Great work to both of you!

**Pull request merged**